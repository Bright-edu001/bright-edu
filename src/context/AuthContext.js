import React, { createContext, useContext, useEffect, useState } from "react";
import {
  getAuth,
  signInWithEmailAndPassword,
  signOut,
  onAuthStateChanged,
  GoogleAuthProvider,
  signInWithPopup,
} from "firebase/auth";
import { app } from "../config/firebaseConfig";
import { message } from "antd";
import {
  getUserRole,
  hasPermission,
  hasAnyPermission,
  getRolePermissions,
  USER_ROLES,
} from "../config/permissions";

const AuthContext = createContext();

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null);
  const [userRole, setUserRole] = useState(null);
  const [userPermissions, setUserPermissions] = useState([]);
  const [loading, setLoading] = useState(true);
  const auth = getAuth(app);

  // 檢查環境
  const isProduction =
    process.env.NODE_ENV === "production" ||
    window.location.hostname !== "localhost";

  // 檢查是否為允許的email網域
  const isAllowedEmail = (email) => {
    return email && email.endsWith("@bright-edu.com");
  };

  // 設定用戶角色和權限
  const setUserRoleAndPermissions = (email) => {
    if (!email) {
      setUserRole(null);
      setUserPermissions([]);
      return;
    }

    const role = getUserRole(email);
    const permissions = getRolePermissions(role);

    setUserRole(role);
    setUserPermissions(permissions);
  };

  // 登入功能
  const login = async (email, password) => {
    try {
      if (isProduction) {
        // 生產環境：使用Firebase Authentication
        if (!isAllowedEmail(email)) {
          throw new Error("只允許 @bright-edu.com 的電子郵件帳號登入");
        }

        const userCredential = await signInWithEmailAndPassword(
          auth,
          email,
          password
        );
        const user = userCredential.user;

        if (!isAllowedEmail(user.email)) {
          await signOut(auth);
          throw new Error("只允許 @bright-edu.com 的電子郵件帳號登入");
        }

        // 設定用戶角色和權限
        setUserRoleAndPermissions(user.email);
        message.success("登入成功！");
        return user;
      } else {
        // 開發環境：使用原來的帳號密碼驗證
        const EXPECTED_USERNAME = "admin";
        const EXPECTED_PASSWORD = "0000";

        if (email === EXPECTED_USERNAME && password === EXPECTED_PASSWORD) {
          // 創建一個模擬的用戶對象
          const mockUser = {
            uid: "dev-user",
            email: "admin@bright-edu.com",
            displayName: "Admin User",
            isDevelopment: true,
          };
          setUser(mockUser);
          // 設定開發環境的超級管理員權限
          setUserRole(USER_ROLES.SUPER_ADMIN);
          setUserPermissions(getRolePermissions(USER_ROLES.SUPER_ADMIN));

          localStorage.setItem("isAuthenticated", "true");
          localStorage.setItem("devUser", JSON.stringify(mockUser));
          message.success("登入成功！");
          return mockUser;
        } else {
          throw new Error("帳號或密碼錯誤！");
        }
      }
    } catch (error) {
      message.error(error.message);
      throw error;
    }
  };

  // Google登入 (僅生產環境)
  const loginWithGoogle = async () => {
    if (!isProduction) {
      message.error("Google登入僅在生產環境可用");
      return;
    }

    try {
      const provider = new GoogleAuthProvider();
      // 強制選擇帳號
      provider.setCustomParameters({
        prompt: "select_account",
      });

      const userCredential = await signInWithPopup(auth, provider);
      const user = userCredential.user;

      if (!isAllowedEmail(user.email)) {
        await signOut(auth);
        throw new Error("只允許 @bright-edu.com 的電子郵件帳號登入");
      }

      // 設定用戶角色和權限
      setUserRoleAndPermissions(user.email);
      message.success("Google登入成功！");
      return user;
    } catch (error) {
      message.error(error.message);
      throw error;
    }
  };

  // 登出功能
  const logout = async () => {
    try {
      if (isProduction && auth.currentUser) {
        await signOut(auth);
      }

      // 清除本地存儲
      localStorage.removeItem("isAuthenticated");
      localStorage.removeItem("devUser");
      setUser(null);
      setUserRole(null);
      setUserPermissions([]);
      message.success("已登出");
    } catch (error) {
      message.error("登出失敗");
      console.error("Logout error:", error);
    }
  };

  // 監聽身份驗證狀態變化
  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (firebaseUser) => {
      if (isProduction) {
        // 生產環境：使用Firebase用戶
        if (firebaseUser && isAllowedEmail(firebaseUser.email)) {
          setUser(firebaseUser);
          setUserRoleAndPermissions(firebaseUser.email);
          localStorage.setItem("isAuthenticated", "true");
        } else {
          setUser(null);
          setUserRole(null);
          setUserPermissions([]);
          localStorage.removeItem("isAuthenticated");
        }
      } else {
        // 開發環境：檢查本地存儲
        const storedAuth = localStorage.getItem("isAuthenticated");
        const storedUser = localStorage.getItem("devUser");

        if (storedAuth === "true" && storedUser) {
          try {
            const parsedUser = JSON.parse(storedUser);
            setUser(parsedUser);
            // 設定開發環境權限
            setUserRole(USER_ROLES.SUPER_ADMIN);
            setUserPermissions(getRolePermissions(USER_ROLES.SUPER_ADMIN));
          } catch (error) {
            console.error("Error parsing stored user:", error);
            localStorage.removeItem("isAuthenticated");
            localStorage.removeItem("devUser");
            setUser(null);
            setUserRole(null);
            setUserPermissions([]);
          }
        } else {
          setUser(null);
          setUserRole(null);
          setUserPermissions([]);
        }
      }

      setLoading(false);
    });

    return () => unsubscribe();
  }, [auth, isProduction]);

  // 檢查是否已登入
  const isAuthenticated = () => {
    return user !== null;
  };

  // 權限檢查函數
  const checkPermission = (permission) => {
    return hasPermission(userRole, permission);
  };

  const checkAnyPermission = (permissionList) => {
    return hasAnyPermission(userRole, permissionList);
  };

  const value = {
    user,
    userRole,
    userPermissions,
    login,
    loginWithGoogle,
    logout,
    isAuthenticated,
    checkPermission,
    checkAnyPermission,
    loading,
    isProduction,
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export default AuthContext;
