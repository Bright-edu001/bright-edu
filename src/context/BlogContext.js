import React, {
  createContext,
  useState,
  useMemo,
  useCallback,
  useEffect,
} from "react";
import { getEnrollmentEvents, getNews } from "../services/blogService";

// 建立 BlogContext，用於全域分享部落格資料與操作方法
export const BlogContext = createContext();

// BlogProvider 包裹元件，提供招生活動、最新消息資料及操作函式
export const BlogProvider = ({ children }) => {
  // 儲存招生活動、最新消息、載入狀態與錯誤訊息
  const [enrollmentEvents, setEnrollmentEvents] = useState([]);
  const [news, setNews] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  // 使用 useEffect 在元件掛載時從 API 獲取資料
  useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        // 同時獲取招生活動與最新消息
        const [enrollmentData, newsData] = await Promise.all([
          getEnrollmentEvents(),
          getNews(),
        ]);
        setEnrollmentEvents(enrollmentData);
        setNews(newsData);
        setError(null); // 成功後清除錯誤
      } catch (err) {
        setError(err.message);
        console.error("Failed to fetch blog data:", err);
      } finally {
        setLoading(false); // 完成後設定為非載入狀態
      }
    };

    fetchData();
  }, []); // 空依賴陣列確保只在掛載時執行一次

  // 使用 useMemo 將招生活動與最新消息合併，避免不必要的重新計算
  const all = useMemo(
    () => [...enrollmentEvents, ...news],
    [enrollmentEvents, news]
  );

  // 根據關鍵字搜尋文章，回傳符合標題或內容包含關鍵字的清單
  const searchByKeyword = useCallback(
    (keyword) => {
      const kw = (keyword || "").trim().toLowerCase();
      if (!kw) return [];
      return all.filter(
        (item) =>
          item.title.toLowerCase().includes(kw) ||
          (item.content && item.content.toLowerCase().includes(kw))
      );
    },
    [all]
  );

  // 根據分類參數過濾文章，支援 "enrollment"、"news" 或回傳全部
  const filterByCategory = useCallback(
    (category) => {
      if (category === "enrollment") return enrollmentEvents;
      if (category === "news") return news;
      return all;
    },
    [enrollmentEvents, news, all]
  );

  // 使用 useMemo 穩定 context value，只在相關數據改變時重新計算
  const contextValue = useMemo(
    () => ({
      enrollmentEvents,
      news,
      all,
      loading, // 新增 loading 狀態
      error, // 新增 error 狀態
      searchByKeyword,
      filterByCategory,
    }),
    [
      enrollmentEvents,
      news,
      all,
      loading,
      error,
      searchByKeyword,
      filterByCategory,
    ]
  );

  // 提供全域資料與操作函式給子元件
  return (
    <BlogContext.Provider value={contextValue}>{children}</BlogContext.Provider>
  );
};
