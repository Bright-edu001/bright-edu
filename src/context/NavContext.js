import React, { createContext, useState, useEffect, useRef, useMemo, useCallback } from "react";

// 建立 NavContext，用於全站導航下拉選單狀態管理
export const NavContext = createContext();

// 全域初始下拉選單狀態，不會於渲染中改變
const initialStates = {
  UIC: false,
  MSU: false,
  UICSub: false,
  MBASub: false,
  MSSub: false,
  chicagoSub: false,
  rankingsSub: false,
  mobileMenu: false,
  areasSub: false,
  msuMainSub: false,
  msfProgramsSub: false,
  eastLansingSub: false,
};

export const NavProvider = ({ children }) => {
  const [dropdownStates, setDropdownStates] = useState(initialStates);
  const [pendingLink, setPendingLink] = useState(null);
  const headerRef = useRef(null);

  // Define the full menu hierarchy
  const fullMenuHierarchy = useMemo(() => ({
    UIC: ["UICSub", "MBASub", "MSSub"],
    MSU: ["msuMainSub", "msfProgramsSub"],
    UICSub: ["chicagoSub", "rankingsSub"],
    MBASub: ["areasSub"],
    MSSub: [],
    msuMainSub: ["eastLansingSub"],
    msfProgramsSub: [],
    chicagoSub: [],
    rankingsSub: [],
    areasSub: [],
    eastLansingSub: [],
    mobileMenu: []
  }), []);

  // Helper to get all descendants (children, grandchildren, etc.)
  const getAllDescendants = useCallback((parentKey, hierarchy) => {
    let descendants = new Set();
    const children = hierarchy[parentKey] || [];
    children.forEach(child => {
      descendants.add(child);
      const grandchildren = getAllDescendants(child, hierarchy);
      grandchildren.forEach(gc => descendants.add(gc));
    });
    return descendants;
  }, []);

  // 點擊選單外關閉所有下拉選單
  useEffect(() => {
    const handleClickOutside = (event) => {
      if (headerRef.current && !headerRef.current.contains(event.target)) {
        setDropdownStates(initialStates);
        setPendingLink(null);
      }
    };
    document.addEventListener("mousedown", handleClickOutside);
    return () => document.removeEventListener("mousedown", handleClickOutside);
  }, []);

  // 通用切換函數
  const toggleDropdown =
    (menuKey, closeOthers = [], childMenus = []) =>
    (e) => {
      if (e) {
        e.preventDefault();
        e.stopPropagation();
      }
      setDropdownStates((prevStates) => {
        const newDropdownStates = { ...prevStates };
        const menuKeyIsCurrentlyOpen = prevStates[menuKey];

        // Toggle the state of the primary menu
        newDropdownStates[menuKey] = !menuKeyIsCurrentlyOpen;

        // Determine which parent menus are being closed (either menuKey itself or those in closeOthers)
        // Their children will also need to be closed.
        const parentMenusBeingClosed = new Set();

        closeOthers.forEach(otherKey => {
          if (newDropdownStates[otherKey] === true || prevStates[otherKey] === true) { // If it was open or is still marked open before this logic
            parentMenusBeingClosed.add(otherKey);
          }
        });

        // If the primary menu (menuKey) was open and is now being closed
        if (menuKeyIsCurrentlyOpen && !newDropdownStates[menuKey]) {
          parentMenusBeingClosed.add(menuKey);
        }
        
        const finalKeysToSetFalse = new Set();
        parentMenusBeingClosed.forEach(parentToClose => {
          finalKeysToSetFalse.add(parentToClose); // Add the parent itself to be closed
          const descendants = getAllDescendants(parentToClose, fullMenuHierarchy);
          descendants.forEach(descendant => finalKeysToSetFalse.add(descendant));
        });

        finalKeysToSetFalse.forEach(keyToClose => {
          // If keyToClose is the menuKey that we just decided to OPEN, don't set it to false.
          if (keyToClose === menuKey && newDropdownStates[menuKey] === true) {
            // Do nothing, keep it open
          } else {
            newDropdownStates[keyToClose] = false;
          }
        });
        
        // If menuKey is being OPENED, ensure its DIRECT children (from childMenus param) are closed (reset).
        if (newDropdownStates[menuKey]) { // If menuKey is now open
          (childMenus || []).forEach(directChildKey => {
            newDropdownStates[directChildKey] = false;
          });
        }
        return newDropdownStates;
      });
    };

  // 定義各選單 toggle
  const toggleUICDropdown = toggleDropdown(
    "UIC",
    ["MSU"],
    ["UICSub", "MBASub", "MSSub", "chicagoSub", "rankingsSub", "areasSub"]
  );
  const toggleMSUDropdown = toggleDropdown(
    "MSU",
    ["UIC"],
    ["msuMainSub", "msfProgramsSub"]
  );
  const toggleUICSubDropdown = toggleDropdown(
    "UICSub",
    ["MBASub", "MSSub"],
    ["chicagoSub", "rankingsSub"]
  );
  const toggleMBASubDropdown = toggleDropdown(
    "MBASub",
    ["UICSub", "MSSub"],
    ["areasSub"]
  );
  const toggleMSSubDropdown = toggleDropdown("MSSub", ["UICSub", "MBASub"]); // Assuming MSSub has no children based on original structure
  const toggleMSUMainSubDropdown = toggleDropdown("msuMainSub", [
    "msfProgramsSub",
  ], ["eastLansingSub"]); // Corrected: Added ["eastLansingSub"] as childMenu
  const toggleMSFProgramsSubDropdown = toggleDropdown("msfProgramsSub", [
    "msuMainSub",
  ]); // Assuming msfProgramsSub has no children
  const toggleMobileMenu = toggleDropdown("mobileMenu");

  // 處理含子選單連結點擊行為
  const handleDropdownLinkClick = useCallback(
    (menuKey, href, closeKey = null) => (e) => {
      const mobileCheck = window.innerWidth < 1025;
      if (mobileCheck) {
        e.preventDefault();
        e.stopPropagation();
        const isOpen = dropdownStates[menuKey];
        const isPending = pendingLink === menuKey;
        if (!isOpen) {
          setDropdownStates((prev) => ({
            ...prev,
            [menuKey]: true,
            ...(closeKey ? { [closeKey]: false } : {}),
          }));
          setPendingLink(menuKey);
        } else if (isPending) {
          setPendingLink(null);
          window.location.href = href;
        } else {
          setPendingLink(menuKey);
        }
      }
    },
    [dropdownStates, pendingLink]
  );

  // 滑鼠懸停展開與離開（桌機）
  const handleMouseEnter = useCallback(
    (menuKey, closeKey) => () => {
      if (window.innerWidth >= 1025) {
        setDropdownStates((prev) => ({
          ...prev,
          [menuKey]: true,
          ...(closeKey ? { [closeKey]: false } : {}),
        }));
        setPendingLink(null);
      }
    },
    []
  );
  const handleMouseLeave = useCallback(
    (menuKey) => () => {
      if (window.innerWidth >= 1025) {
        setDropdownStates((prev) => ({ ...prev, [menuKey]: false }));
        setPendingLink(null);
      }
    },
    []
  );

  // 使用 useMemo 穩定 context value
  const contextValue = useMemo(
    () => ({
      dropdownStates,
      toggleUICDropdown,
      toggleMSUDropdown,
      toggleUICSubDropdown,
      toggleMBASubDropdown,
      toggleMSSubDropdown,
      toggleMSUMainSubDropdown,
      toggleMSFProgramsSubDropdown,
      toggleMobileMenu,
      handleDropdownLinkClick,
      handleMouseEnter,
      handleMouseLeave,
      headerRef,
    }),
    [
      dropdownStates,
      toggleUICDropdown,
      toggleMSUDropdown,
      toggleUICSubDropdown,
      toggleMBASubDropdown,
      toggleMSSubDropdown,
      toggleMSUMainSubDropdown,
      toggleMSFProgramsSubDropdown,
      toggleMobileMenu,
      handleDropdownLinkClick,
      handleMouseEnter,
      handleMouseLeave,
      headerRef,
    ]
  );
  return (
    <NavContext.Provider value={contextValue}>
      {children}
    </NavContext.Provider>
  );
};
