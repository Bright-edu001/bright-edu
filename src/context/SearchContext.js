import React, { createContext, useState } from "react";
import { useNavigate } from "react-router-dom";

// 建立搜尋上下文，管理全站關鍵字與分類導向功能
export const SearchContext = createContext();

export const SearchProvider = ({ children }) => {
  // 關鍵字狀態
  const [keyword, setKeyword] = useState("");
  const navigate = useNavigate();

  // 處理關鍵字搜尋表單提交，導向搜尋結果頁面並清空輸入
  const handleSearch = (e) => {
    e.preventDefault();
    const kw = keyword.trim();
    if (kw) {
      navigate(`/blog/search/${encodeURIComponent(kw)}`);
      setKeyword("");
    }
  };

  // 處理分類點擊，導向對應分類列表
  const handleCategoryClick = (category) => {
    if (category === "enrollment") {
      navigate("/blog?category=enrollment");
    } else if (category === "news") {
      navigate("/blog?category=news");
    }
  };

  return (
    <SearchContext.Provider
      value={{ keyword, setKeyword, handleSearch, handleCategoryClick }}
    >
      {children}
    </SearchContext.Provider>
  );
};
