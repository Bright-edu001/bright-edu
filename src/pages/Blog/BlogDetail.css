.responsive-img{width:100%;-o-object-fit:contain;object-fit:contain;display:block}.blog-detail-mainrow{display:flex;gap:2rem;margin-top:5rem;padding:0}.blog-detail-main{display:flex;flex-direction:column;min-width:0;padding:0}.blog-detail-categories h3{font-size:32px;margin-bottom:1rem;font-weight:900;color:#1c184a}.blog-detail-categories ul{list-style:none;font-size:20px;padding:0;margin:0}.blog-detail-categories li+li{margin-top:.5rem;font-weight:700}.blog-detail-category-btn{background:none;border:none;color:#1c184a;font-size:20px;font-weight:700;cursor:pointer;padding:.3rem .5rem;border-radius:4px}.blog-detail-category-label{font-size:1.5rem;color:#1c184a;font-weight:500;letter-spacing:.5px;transition:color .1s ease-in-out;cursor:pointer}.blog-detail-category-label:hover{color:#c71432;text-decoration:underline}.blog-detail-image{width:1000px;padding:0 1.5rem;-o-object-fit:cover;object-fit:cover;max-width:100%;margin-bottom:1rem}.blog-detail-back{display:flex;justify-content:center;font-size:24px;color:#1c184a;margin-top:2rem;margin-bottom:1rem}.blog-detail-back:hover{color:#c71432;transform:scale(1.05);transition:all .1s ease-in-out}.blog-detail-notfound{padding:2rem}.blog-detail-content p{margin-top:1.2em;margin-bottom:1.2em}@media screen and (max-width: 850px){.blog-detail-mainrow{flex-direction:column-reverse;gap:1.5rem;margin-top:2rem}.blog-detail-excerpt{margin-top:2rem}.blog-detail-image{width:100%;height:200px;max-width:100%;margin-bottom:1rem;margin-left:0}.blog-detail-back{margin-left:0;font-size:20px}}/*# sourceMappingURL=BlogDetail.css.map */