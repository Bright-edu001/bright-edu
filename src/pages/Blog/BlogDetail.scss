@import "../../styles/variables";
@import "../../styles/mixins";

// 主內容與側邊欄橫向排列
.blog-detail-mainrow {
  display: flex;
  gap: 2rem;
  margin-top: 4rem;
  margin-bottom: 7rem;
  padding: 0;
  justify-content: center;
  @media screen and (max-width: 1025px) {
    align-items: normal;
  }
}

// 主內容
.blog-detail-main {
  display: flex;
  flex-direction: column;
  min-width: 0;

  .blog-detail-title {
    font-size: 24px;
    font-weight: 700;
    text-align: left;
    padding: 0 1.5rem;
  }
  .blog-detail-excerpt {
    font-size: 18px;
    text-align: left;
    padding: 0 1.5rem;
    @media screen and (max-width: 500px) {
      margin: 1rem 0;
    }
  }
  .blog-detail-category-label {
    font-size: 24px;
    font-weight: 700;
    color: $color-nav;
    text-align: left;
    padding: 0 1.5rem;
    transition: color 0.1s ease-in-out;
    cursor: pointer;
    &:hover {
      color: $color-uic;
      text-decoration: underline;
    }
  }
  .blog-detail-content {
    font-size: 18px;
    text-align: left;
    padding: 0 1.5rem;
    color: $color-black;

    .detail-item {
      // 預設無縮排
      &.nested {
        margin-left: 20px; // 套用至巢狀項目
        @media screen and (max-width: 500px) {
          margin-left: 0; // 手機版縮小縮排
          ul {
            display: flex;
            flex-direction: column;
            list-style-position: inside;
            gap: $spacing-xs;
          }
        }
      }
    }

    .detail-icon {
      margin-right: 10px;
    }
  }
}

.blog-detail-categories {
  h3 {
    font-size: 24px;
    margin-bottom: 1rem;
    font-weight: 700;
    color: $color-nav;
  }
  ul {
    list-style: none;
    font-size: 18px;
    padding: 0;
    margin: 0;
  }
  li + li {
    margin-top: 0.5rem;
    font-weight: 700;
  }
}

.blog-detail-category-btn {
  background: none;
  border: none;
  color: $color-nav;
  font-size: $font-size-lg;
  font-weight: 700;
  cursor: pointer;
  padding: 0.3rem 0.5rem;
  border-radius: 4px;
}

.blog-detail-image {
  width: 1000px;
  padding: 0 1.5rem;
  object-fit: cover;
  max-width: 100%;
  margin-bottom: 1rem;
}

.blog-detail-back {
  .blog-back-btn {
    margin: 3rem auto;
    display: block;
    padding: 10px 24px;
    font-size: 1rem;
    border-radius: 8px;
    border: 1px solid #ccc;
    background: #fff;
    cursor: pointer;
    transition: background 0.2s, color 0.2s, border 0.2s;
    color: $color-black;
    &:hover {
      background: $color-nav;
      color: $color-white;
      border-color: $color-nav;
    }
  }
}

.blog-detail-notfound {
  padding: 2rem;
}

.blog-detail-content p {
  margin-top: 1.2em;
  margin-bottom: 1.2em;
}

// 響應式設計：寬度850px以下
@media screen and (max-width: 850px) {
  .blog-detail-mainrow {
    flex-direction: column;
    gap: 1.5rem;
    margin-top: $spacing-lg;
    align-items: normal;
  }

  .blog-detail-excerpt {
    margin-top: 2rem;
  }
  .blog-detail-image {
    width: 100%;
    height: 200px;
    max-width: 100%;
    margin-bottom: 1rem;
    margin-left: 0;
  }

  .blog-detail-back {
    margin-left: 0;
    font-size: $font-size-lg;
  }
}
