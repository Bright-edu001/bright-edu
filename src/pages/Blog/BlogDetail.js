import React, { useContext } from "react";
import "./BlogDetail.scss";
import { useParams, Link } from "react-router-dom";
import { BlogContext } from "../../context/BlogContext";
import MbaAreasHero from "../../components/MbaAreasHero/MbaAreasHero";
import SearchBar from "../../components/SearchBar/SearchBar";
import "../../styles/critical.css";
import sanitizeHtml from "../../utils/sanitizeHtml";

const hashText = (text = "") => {
  let hash = 0;
  for (let i = 0; i < text.length; i++) {
    hash = (hash << 5) - hash + text.charCodeAt(i);
    hash |= 0;
  }
  return hash.toString();
};

// 遞迴渲染詳細資訊的函式
const renderDetails = (details, isNested = false) => {
  return details.map((detail) => {
    const key = detail.id || hashText(detail.text || JSON.stringify(detail));
    return (
      <div key={key} className={`detail-item ${isNested ? "nested" : ""}`}>
        <p>
          {detail.icon && <span className="detail-icon">{detail.icon}</span>}
          {detail.text}
        </p>
        {detail.list && (
          <ul>
            {detail.list.map((item) => (
              <li key={hashText(item)}>{item}</li>
            ))}
          </ul>
        )}
        {detail.subDetails && renderDetails(detail.subDetails, true)}
      </div>
    );
  });
};

function BlogDetail() {
  const { id } = useParams();
  const { enrollmentEvents, news, all, loading, error } =
    useContext(BlogContext);
  const blog = all.find((item) => String(item.id) === String(id));

  // 新增：處理載入與錯誤狀態
  if (loading) {
    return <div>載入中...</div>;
  }

  if (error) {
    return <div>讀取文章時發生錯誤: {error}</div>;
  }

  // 新增：判斷分類
  let subtitle = "";
  if (blog) {
    if (enrollmentEvents.some((item) => String(item.id) === String(blog.id))) {
      subtitle = "招生活動";
    } else if (news.some((item) => String(item.id) === String(blog.id))) {
      subtitle = "最新消息";
    }
  }

  // 計算分類參數用於Link
  const categoryParam = subtitle === "招生活動" ? "enrollment" : "news";

  // 搜尋與分類按鈕導向由 SearchContext handleSearch, handleCategoryClick 管理

  if (!blog) {
    return <div>找不到文章</div>;
  }

  return (
    <div>
      <MbaAreasHero />
      <div className="blog-detail-mainrow">
        <div className="blog-detail-main">
          <img
            className="blog-detail-image"
            src={blog.image}
            alt={blog.title}
          />
          <h1 className="blog-detail-title emoji-support">{blog.title}</h1>
          {/* 分類標籤：改用Link以提供href */}
          <Link
            className="blog-detail-category-label"
            to={`/blog?category=${categoryParam}`}
            title={`查看${subtitle}分類`}
          >
            {subtitle}
          </Link>
          <p className="blog-detail-excerpt">{blog.excerpt}</p>
          <div className="blog-detail-content">
            {typeof blog.content === "string"
              ? blog.content
                  .split("\n")
                  .filter((line) => line.trim() !== "")
                  .map((line, idx) => <p key={idx}>{line}</p>)
              : blog.type === "enrollment"
              ? blog.content.map((semesterInfo, index) => (
                  <div key={index} className="semester-section">
                    <h2
                      className="emoji-support"
                      dangerouslySetInnerHTML={{
                        __html: sanitizeHtml(semesterInfo.title),
                      }}
                    />
                    {renderDetails(semesterInfo.details)}
                  </div>
                ))
              : blog.type === "article" && typeof blog.content === "object"
              ? renderDetails(blog.content.details)
              : null}
          </div>
          <Link to="/blog" className="blog-back-btn blog-detail-back">
            ← 返回部落格
          </Link>
          {/* 這裡可根據需求擴充更多內容 */}
        </div>
        <aside className="blog-detail-sidebar">
          {/* 關鍵字搜尋欄 */}
          <SearchBar placeholder="搜尋..." />
        </aside>
      </div>
    </div>
  );
}

export default BlogDetail;
