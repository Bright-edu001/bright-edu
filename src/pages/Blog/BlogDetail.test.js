import React from "react";
import "@testing-library/jest-dom";
import { render, screen, within } from "@testing-library/react";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import BlogDetail from "./BlogDetail";
import { BlogContext } from "../../context/BlogContext";

jest.mock("../../components/MbaAreasHero/MbaAreasHero", () => () => <div />);
jest.mock("../../components/SearchBar/SearchBar", () => () => <div />);

const providerValue = {
  enrollmentEvents: [{ id: 1 }],
  news: [],
  all: [
    {
      id: 1,
      title: "Sample",
      image: "img.jpg",
      excerpt: "excerpt",
      type: "enrollment",
      content: [
        {
          title: 'Hello<script>alert("xss")</script>',
          flagImage: "/flag.png",
          details: [],
        },
      ],
    },
  ],
  loading: false,
  error: null,
};

const wrapper = ({ children }) => (
  <BlogContext.Provider value={providerValue}>
    <MemoryRouter initialEntries={["/blog/1"]}>
      <Routes>
        <Route path="/blog/:id" element={children} />
      </Routes>
    </MemoryRouter>
  </BlogContext.Provider>
);

describe("BlogDetail", () => {
  test("renders flag image and escapes title HTML", () => {
    render(<BlogDetail />, { wrapper });
    const heading = screen.getByRole("heading", { level: 2 });
    const img = within(heading).getByRole("img");
    expect(img).toHaveAttribute("src", "/flag.png");
    expect(heading.innerHTML).toContain(
      '&lt;script&gt;alert("xss")&lt;/script&gt;'
    );
    expect(heading.innerHTML).not.toContain("<script>");
  });

  test("renders stably across rerenders", () => {
    const { asFragment, rerender } = render(<BlogDetail />, { wrapper });
    // 使用 snapshot 驗證重渲染前後輸出一致，避免直接存取 DOM 節點
    rerender(<BlogDetail />);
    expect(asFragment()).toMatchSnapshot();
  });
});
