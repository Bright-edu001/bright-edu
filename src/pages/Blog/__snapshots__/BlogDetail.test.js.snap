// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogDetail renders stably across rerenders 1`] = `
<DocumentFragment>
  <div>
    <div />
    <div
      class="blog-detail-mainrow"
    >
      <div
        class="blog-detail-main"
      >
        <img
          alt="Sample"
          class="blog-detail-image"
          height="571"
          src="img.jpg"
          width="1000"
        />
        <h1
          class="blog-detail-title emoji-support"
        >
          Sample
        </h1>
        <a
          class="blog-detail-category-label"
          href="/blog?category=enrollment"
          title="查看招生活動分類"
        >
          招生活動
        </a>
        <p
          class="blog-detail-excerpt"
        >
          excerpt
        </p>
        <div
          class="blog-detail-content"
        >
          <div
            class="semester-section"
          >
            <h2
              class="emoji-support"
            >
              <img
                alt="flag"
                class="flag-icon"
                loading="lazy"
                src="/flag.png"
              />
              Hello&lt;script&gt;alert("xss")&lt;/script&gt;
            </h2>
          </div>
        </div>
        <a
          class="blog-back-btn blog-detail-back"
          href="/blog"
        >
          ← 返回部落格
        </a>
      </div>
      <aside
        class="blog-detail-sidebar"
      >
        <div />
      </aside>
    </div>
  </div>
</DocumentFragment>
`;
